### ISSUE TITLE
RFC: Drop Ruby 2.2 support for Puma 5.0

### ISSUE THREAD

**Issue by nateberkopec** [puma/puma#2245]
👋 Hola Puma community.

We (active commit bit group: myself, @schneems, @evanphx) are considering dropping Ruby 2.2 support for Puma 5.0.

Here's why:

* We're looking at the public numbers on Rubygems and seeing >80% of active gem installers are using Ruby 2.3+.
* Based on our own sense of the community, we think that _at least_ that number above are using Ruby 2.3+.
* Ruby 2.2 is hard to install nowadays, especially re: SSL.
* We have a number of hacks in the CI, tests and lib that are accomodations to make Ruby 2.2 work.
* Ruby 2.2 never fixed the thread interrupt queue bug, so technically to use Puma with Ruby 2.2 you should be using the hotfix gem.

Puma's ruby version policy is more or less that we want to support as many active Ruby web applications as possible, and view ourselves somewhere closer to "deep infrastructure" like Rack than "let's push everyone's versions forward" like Rails or pegging our support policy directly to Ruby's EOL policy.

**If think there's a good reason that Puma should support Ruby 2.2, speak now or forever hold your peace.** This thread is not for discussion about dropping higher versions of Ruby. Only 2.2 is on the table.

**Comment by nateberkopec** [comment]
Also to clarify: Puma 5.0 is _the next release_, which will occur as soon as everything [in this milestone is cleared out.](https://github.com/puma/puma/milestone/7)

**Comment by ioquatix** [comment]
I don't know if my opinion holds much weight but I'd say, at the point of a major release, you should only bother suppporting non-EOL Rubies. People who need to support older rubies can just use older releases that still work. You, as an open source maintainer are under no obligation to support businesses who cannot invest the bare minimum of effort to keep things up to date IMHO, and the cost of supporting old rubies is not insignificant.

**Comment by nateberkopec** [comment]
@MSP-Greg brought up a good point offline that we're very close to 5.0 today and currently CI passes on 2.2. So we may just remove 2.2 support on 6.0 instead.

**Comment by rafbm** [comment]
Perhaps just to rephrase @ioquatix’s thoughts more bluntly, I don’t see under what circumstances anyone could reasonably consider upgrading Puma from 4.x to 5.x a higher priority than upgrading Ruby from 2.2 to 2.[3,4,5]. Genuinely curious to hear a good argument for this.

If a security issue is found in your 4.x branch, [you would backport](https://github.com/puma/puma/security/policy) it even after 5.x is released, correct? That rules out the security argument.

Is it because you want to avoid people doing `bundle update` blindly and breaking their app? `required_ruby_version` has you covered.

**Comment by ioquatix** [comment]
> @MSP-Greg brought up a good point offline that we're very close to 5.0 today and currently CI passes on 2.2. So we may just remove 2.2 support on 6.0 instead.

I could also add, that I don't see this as a reason to support 2.2 - just because it works *now* doesn't mean you want to support it for the life time of 5.x.

**Comment by MSP-Greg** [comment]
Ok.  To clarify:

1. There are lot of bug fixes in History.md.
2. There are some items in History's 'Deprecations, Removals and Breaking API Changes'.  Hence, releasing as 4.x would be improper.
3. CI is passing on Ruby 2.2.

Given the large number of bug fixes, release 5.0.0, and announce that work is beginning on 6.0, with the intent to drop Ruby 2.2 support.  Given the similarity between 2.2 and 2.3, I think dropping both should be considered for 6.0.  Given that OpenSSL 1.0.2 is EOL, I don't feel dropping 2.3 is drastic.

As to whether bugs will be backported to 5.0.x, that's the maintainers' decision.

**Comment by danmayer** [comment]
I have been struggling with exact releases to support as well on a much smaller userbase with gems... I was originally trying to follow heroku and got a bit of pushback... I did drop 2.2 with no complaints, and I am supporting 2.3 only because we have no valid breaking changes yet.

Anyways, getting a bit more of a community standard would be good... something like LTS, where we support Ruby EOL +1 or something.

**Comment by dentarg** [comment]
I'm in favour of dropping Ruby 2.2 support. 👍 IMHO, it doesn't make sense as an open-source project to spend our scarce resources on legacy versions.

As a reference, [rack dropped 2.2 support in version 2.2.0 (2020-02-08)](https://github.com/rack/rack/blob/v2.2.2/CHANGELOG.md#220---2020-02-08) (and are discussing the future in https://github.com/rack/rack/issues/1601).

**Comment by MSP-Greg** [comment]
As I stated, there are a lot of bug fixes that have not been released.  Using HISTORY.md, a quick count showed that they equal all the bugs fixes in the releases going back maybe seven months?

I've always been a proponent of using the latest release of everything, and I think I've shown that by my 'encouraging' the use of Ruby master builds.  Some of my work with Puma has involved OpenSSL, and I would love to remove support of OpenSSL 1.0.2...

Regardless, to have a large number of bug fixes that work with Ruby 2.2 and not release them seems inappropriate.  Since there are breaking changes, release as 5.0, and start 6.0.

Last post re this issue, I thought I'd clearly state a few things.

**Comment by kingdonb** [comment]
I would suggest that if CI passes on Ruby 2.2, great, but as you have made advance notice and appear to have planned to drop support, I'd say you should consider dropping support altogether as soon as it becomes an obstacle to ongoing maintenance.

The RSpec 4 (rspec-rails) release story is a good example of what can happen when you support too-old versions for too long. They have the difficult situation of being beholden to two upstreams with fairly different support timelines (that is, Ruby and Rails)

You can "soft-deprecate" Ruby 2.2 with a warning message to help people think about upgrading soon, which they should be doing if they want to remain in support anyway, then perhaps on a minor release later, whenever it has become too much to keep supporting the older version, you can add the version lock and finalize the hard deprecation.

It doesn't need to be hard-deprecated as soon as the first minor release in 5.x, but if you add the warning now, then you know you may opt to make the cut at any point going forward, and nobody can act surprised about it.

I am one who is still struggling to upgrade some apps from 2.0, and as a maintainer who answers to management, I have to say I would really appreciate it if more gems added hard deprecation for old and unsupported releases, so we would have more reasons to offer to justify why upgrading is needed. "Look, this x, y, and z dependency that we use are all saying it's time to upgrade. Don't we trust them?"

The boss does not care about new language features that might help us work faster as developers (they should), but another vendor/upstream who says we can no longer get security patches applied without upgrading our core language, may catch the attention of those who can help us force the issue. I mean, by forcing the issue, to help us make good decisions when it comes time to insist on doing upgrades for apps that are kept online, but in maintenance mode.

**Comment by schneems** [comment]
This is an extremely complicated subject. With many grey areas. There's two topics here. One is a general "gems should do <this> as a policy" and the second is "puma's specific case should do <this>"

- Puma doesn't support multiple releases right now. We don't have that kind of bandwidth, so the latest is the only thing we release bugfixes for. We talked about forking version releases and decided against it previously.

I care about maintainer burden, both for puma and for upstream/downstream maintainers. What we choose to do here can and will have unforeseen consequences. One example is someone using puma to benchmark multiple versions of Ruby, if we limit the version then their benchmarks are also limited. Not saying this is a use case we should support, but it is a consequence of limiting the version. 

I think cutting off support for a Ruby version before it's EOL is not reasonable for most libraries, but I can certainly understand if some recent library comes out and wants to use the latest version features. 

I recently released a new gem https://github.com/zombocom/rate_throttle_client that I intend to  be used by all Heroku-16+ customers (if they want/need it). So I made sure it's backward compatible to 2.2 and that's running on CI (via circle-ci). 

All that being said, I think each needs a case-by-case evaluation. 

- What are the benefits from removing 2.2 support in Puma?
  - We can remove some hacks
  - We can use some newer language features such as `+""` for unfrozen strings

I actually previously rev-ed the Ruby version to 2.5 before reverting it. https://github.com/puma/puma/pull/1813

We won't get the nice, fast `.match?` but I don't think that's on any hot paths.

All-in-all I'm in favor of dropping 2.2.
